ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,858 @@
+* if the job is completed with variables then these variables will be stored in the state until they are used for the output mapping (i.e. temporal/transient variables). They must therefore also be written into the `Job:Completed` event.
+* if the service task has an output variable mapping, or the job is completed with variables, then it produces one or more `Variable:Created`/`Variable:Updated` events for the new and updated variables of the workflow instance or scope
+* the order of writing follow-up events matters. Both `ServiceTask:Element_Activated` and `Variable:Created` need to be replayed before `Job:Created`.
+
+On processing:
+* the `ServiceTask:Element_Activating` command transitions the service task
+  element to activated, may create variables for input mapping and creates a job
+* the `JobBatch:Activate` command activates a batch of jobs
+* the `Job:Complete` command completes a job
+* the `ServiceTask:Element_Completing` command transitions the service task
+  element to completed, may create and/or update variables for output mapping
+  and marks the sequence flow as taken
+
+On replay:
+* the `ServiceTask:Element_Activated` event transitions the service task element
+  to activated
+* the `Variable:Created` events create a variables for input mapping
+* the `Job:Created` event creates a job
+* the `JobBatch:Activated` event activates a batch of jobs
+* the `Job:Completed` event completes a job
+* the `Variable:Created` events create variables for output mapping
+* the `Variable:Updated` events update variables for output mapping
+* the `ServiceTask:Element_Completed` event transitions the service task element
+  to completed
+* the `Sequence_Flow_Taken` event marks the sequence flow as taken
+* processing can continue again at `EndEvent:Element_Activating`
+
+Differences to current processing:
+* the variables a job is completed with, are currently not written to the log
+  stream
+
+Considered alternatives:
+* just like call activity variable propagation we could consider storing job
+  variables in `TemporaryVariable:Created` events
+
+### Message Intermediate Catch Event
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+IntermediateCatchEvent:Element_Activating | IntermediateCatchEvent:Element_Activated, MessageSubscription:Open
+MessageSubscription:Open | MessageSubscription:Opened, WorkflowInstanceSubscription:Open
+WorkflowInstanceSubscription:Open | WorkflowInstanceSubscription:Opened
+Message:Publish | Message:Published, WorkflowInstanceSubscription:Correlate
+WorkflowInstanceSubscription:Correlate | WorkflowInstanceSubscription:Correlated, MessageSubscription:Correlate, IntermediateCatchEvent:Element_Completing
+MessageSubscription:Correlate | MessageSubscription:Correlated
+IntermediateCatchEvent:Element_Completing | [Variable:Created], IntermediateCatchEvent:Element_Completed, Sequence_Flow_Taken, EndEvent:Element_Activating
```

Comment:
Let me see if I understand how this would work with two partitions, 1 and 2. 1 has a message subscription opened for message M. 2 has a workflow instance subscription for message M.

I publish message M - this writes `Message:Publish` on 1.
Process `Message:Publish` on 1 - this writes `Message:Published` on 1, and sends the command `WorkflowInstanceSubscription:Correlate` to 2 (but that's not guaranteed to be written).

Now, if we crashed, then replayed from here, this holds because the subscription is in state `correlating`, so we will retry it after some timeout on recovery (or do we immediately retry them after recovery? no matter, we retry them).

In either case, `WorkflowInstanceSubscription:Correlate` is written on 2.
`WorkflowInstanceSubscription:Correlate` is processed on 2 - we write `WorkflowInstanceSubscription:Correlated` on 2, and send `MessageSubscription:Correlate` back to 1, and write `IntermediateCatchEvent:Element_Completing` on 2.

Now if we crash here, `MessageSubscription:Correlate` is not retried, correct? So 1 does not know that about the correlation, and will keep retrying it. Also, if the correlation is rejected, what happens? Would 1 still retry? There would be no `MessageSubscription:RejectCorrelate` command to send back or anything, right?

But let's say 1 doesn't get `MessageSubscription:Correlate`. Then we still complete, and we would send `MessageSubscription:Close`? I don't see here any closing of the subscriptions. Was this omitted for brevity?

Reply:
@npepinpe You've got a good grasp on the topic. I've tried to answer your questions.

@saig0 please correct me if I'm wrong here.

> Now if we crash here, MessageSubscription:Correlate is not retried, correct?

That is correct, but I discussed this previously with @saig0 and we concluded that it appears the current engine is also unable to deal with this.

> Also, if the correlation is rejected, what happens?

This is a great question and had not yet been documented. Correlations should be rejected during processing of `Terminate_Element` command (for example when the workflow instance is cancelled concurrently) or during processing of `Complete_Element` command (for example for a message boundary event on a service task where the job is completed just before the message is published). Currently this happens on processing the `Element_Completing` and `Element_Terminating` records.

I'll add an additional example for this.

> I don't see here any closing of the subscriptions.

The subscription would be closed with `MessageSubscription:Close` command in the case of a rejection. This will be added in the new example